components/scream/src/physics/p3/p3_functions_f90.cpp

```diff
@@ -0,0 +1,188 @@
+
+//
+// A C++ interface to micro_p3 fortran calls and vice versa
+//
+
+
+extern "C" {
+
+void p3_init_a_c(Real* itab, Real* itabcol);
+
+void find_lookuptable_indices_1a_c(Int* dumi, Int* dumjj, Int* dumii, Int* dumzz,
+                                   Real* dum1, Real* dum4, Real* dum5, Real* dum6,
+                                   Real qitot, Real nitot, Real qirim, Real rhop);
+
+void find_lookuptable_indices_1b_c(Int* dumj, Real* dum3, Real qr, Real nr);
+
+void access_lookup_table_c(Int dumjj, Int dumii, Int dumi, Int index,
+                           Real dum1, Real dum4, Real dum5, Real* proc);
+
+void access_lookup_table_coll_c(Int dumjj, Int dumii, Int dumj, Int dumi, Int index,
+                                Real dum1, Real dum3, Real dum4, Real dum5, Real* proc);
+
+}
+
+namespace scream {
+namespace p3 {
+
+//
+// In all C++ -> Fortran bridge functions you should see p3_init(true). P3 needs
+// to be initialized since most of its function depend on global tables to be
+// populated. The 'true' argument is to set p3 to use its fortran implementations
+// instead of calling back to C++. We want this behavior since it doesn't make much
+// sense for C++ to bridge over to fortran only to have fortran bridge back to C++.
+// If the client wanted the C++ implementation, they should just call it directly.
+//
+
+void p3_init_a(P3InitAFortranData& d)
+{
+  p3_init(true); // need to initialize p3 first so that tables are loaded
+  p3_init_a_c(d.itab.data(), d.itabcol.data());
+}
+
+void find_lookuptable_indices_1a(LookupIceData& d)
+{
+  p3_init(true); // need to initialize p3 first so that tables are loaded
+  find_lookuptable_indices_1a_c(&d.dumi, &d.dumjj, &d.dumii, &d.dumzz,
+                                &d.dum1, &d.dum4, &d.dum5, &d.dum6,
+                                d.qitot, d.nitot, d.qirim, d.rhop);
+}
+
+void find_lookuptable_indices_1b(LookupIceDataB& d)
+{
+  p3_init(true);
+  find_lookuptable_indices_1b_c(&d.dumj, &d.dum3, d.qr, d.nr);
+}
+
+void access_lookup_table(AccessLookupTableData& d)
+{
+  p3_init(true); // need to initialize p3 first so that tables are loaded
+  access_lookup_table_c(d.lid.dumjj, d.lid.dumii, d.lid.dumi, d.index,
+                        d.lid.dum1, d.lid.dum4, d.lid.dum5, &d.proc);
+}
+
+void access_lookup_table_coll(AccessLookupTableCollData& d)
+{
+  p3_init(true);
+  access_lookup_table_coll_c(d.lid.dumjj, d.lid.dumii, d.lidb.dumj, d.lid.dumi, d.index,
+                             d.lid.dum1, d.lidb.dum3, d.lid.dum4, d.lid.dum5, &d.proc);
+}
+
+P3GlobalForFortran::P3GlobalForFortran() :
+  itab(new view_itab_table()), itabcol(new view_itabcol_table())
+{
+  P3F::init_kokkos_ice_lookup_tables(*itab, *itabcol);
+}
+
+const P3GlobalForFortran& P3GlobalForFortran::get()
+{
+  static P3GlobalForFortran inst;
+  return inst;
+}
+
+void P3GlobalForFortran::deinit()
+{
+  P3GlobalForFortran& inst = const_cast<P3GlobalForFortran&>(P3GlobalForFortran::get());
+  delete inst.itab;
```

Comment:
I think itab and itabcol should be shared_ptrs, not raw pointers. But would you explain why they need to be pointers at all?

Reply:
If I recall, I was having lots of problems with Kokkos views being deleted after Kokkos finalize. The shared_ptr documentation seemed to indicate that the destructor should be called immediately upon loss of the last pointer, but that didn't seem to be the case in my testing.

Reply:
Static data in function scope get destroyed after main exits, which is the problem re: views you report.

Separate from that, it's better to funnel through one explicit destruction. In the current code, if you were to add another member view pointer, you'd have to remember to delete it. Instead, I recommend you make a static shared_ptr<P3GlobalForFortran> inst, initialize it on the first call, and set it to nullptr in deinit. The internal data can then be plain Views; these will be destroyed when inst is set to nullptr. Make inst private static data in P3GlobalForFortran so deinit can mod the shared_ptr.

Reply:
@ambrad , I will give it another try with the approach you suggested. Just to be clear, I tried having the views as members of P3Global, which failed as expected for the reason you described. I also tried having them as shared_ptrs, with deinit setting them to nullptr. To my surprise, this did not appear to trigger the destructor in time before kokkos finalize.

Reply:
If you find that setting a shared_ptr to nullptr doesn't work as expected, try this standalone reproducer:
```
#include <memory>
#include <iostream>

struct Foo {
  using Ptr = std::shared_ptr<Foo>;

  Foo () { std::cout << "Foo\n"; }
  ~Foo () { std::cout << "~Foo\n"; }

  static const Foo& get () {
    if (inst == nullptr) inst = std::make_shared<Foo>();
    return *inst;
  }

  static void deinit () { inst = nullptr; }

private:
  static Ptr inst;
};

Foo::Ptr Foo::inst;

int main () {
  std::cout << "main start\n";
  const auto& f = Foo::get();
  Foo::deinit();
  std::cout << "main end\n";
}
```
and let me know if you don't see
```
main start
Foo
~Foo
main end
```

Reply:
@ambrad , apologies for the confusion. I was not calling deinit in the right spot, so the only reason dumb pointers were working (and shared_ptrs were not) is that the View was being leaked.

Reply:
@ambrad see if you like this current implementation better.